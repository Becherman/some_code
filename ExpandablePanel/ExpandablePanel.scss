$mobilePadding: 1rem;
$mainPadding: 1.5rem;
$transitionSpeed: 300ms;

:local(.wrapper) {
  background: $white;
  box-shadow: $box-shadow-1;
}

:local(.header) {
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding: $mobilePadding;
  cursor: pointer;

  &:hover {
    & :local(.trigger) {
      color: $brand-primary;
    }
  }

  @include for-tablet {
    @include font-title;
    padding: $mainPadding;
  }
}

:local(.iconContainer) {
  display: flex;
  align-items: center;
}

:local(.content) {
  background: $white;
  overflow: hidden;
  transition: height $transitionSpeed linear 0s;
}

:local(.trigger) {
  cursor: pointer;
  color: $gray-400;
  font-size: $icon-cevron-size;
  transition: all $transitionSpeed linear 0s;
  transform: rotateZ(0);
}

:local(.isOpened) {
  transform: rotateZ(180deg);
}

:local(.text) {
  color: $text-main3;

  @include font-subheader;
  @include for-tablet {
    @include font-title;
  }
}
:local(.icon) {
  color: $brand-primary;
  margin-right: 12px;
  font-size: 28px;
}
