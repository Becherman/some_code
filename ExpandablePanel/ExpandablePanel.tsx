import * as React from 'react';
import cn from 'classnames';
import styles from './ExpandablePanel.scss';

type CSSClasses = 'wrapper' | 'header' | 'content';
type Classes = Partial<Record<CSSClasses, string>>;

interface Props {
  text: string | React.ReactNode;
  children: React.ReactNode;

  defaultExpanded?: boolean;
  expanded?: boolean;

  icon?: string;
  iconRenderer?: React.ReactNode;

  hasTrigger?: boolean;
  triggerRenderer?: React.ReactNode;

  classes?: Classes;

  asyncContent?: boolean;

  onChange?: (event: React.MouseEvent, expanded: boolean) => void;
}

interface State {
  isExpanded: boolean;
  contentHeight: number | string;
}

export class ExpandablePanel extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      isExpanded: !!props.defaultExpanded || !!props.expanded,
      contentHeight: props.defaultExpanded ? 'auto' : 0,
    };
  }

  private content = React.createRef<HTMLDivElement>();

  componentDidUpdate(prevProps: Props, prevState: State) {
    if (
      prevState.contentHeight === 'auto' &&
      this.state.contentHeight !== 'auto'
    ) {
      setTimeout(() => this.setState({ contentHeight: 0 }), 1);
    }
  }

  clickHandler = (event: React.MouseEvent) => {
    const { onChange, expanded } = this.props;
    this.setState(
      state => ({
        isExpanded: expanded === undefined ? !state.isExpanded : expanded,
        contentHeight:
          this.content.current && this.content.current.scrollHeight,
      }),
      () => onChange && onChange(event, this.state.isExpanded),
    );
  };

  updateAfterTransition = () => {
    if (this.state.isExpanded) {
      this.setState({ contentHeight: 'auto' });
    }
  };

  render() {
    const {
      children,
      icon,
      iconRenderer,
      text,
      triggerRenderer,
      classes = {},
      hasTrigger = true,
    } = this.props;
    const { isExpanded, contentHeight } = this.state;
    const { wrapper, header, content } = classes;

    return (
      <div className={cn(styles.wrapper, { [wrapper]: wrapper })}>
        <div
          className={cn(styles.header, { [header]: header })}
          onClick={this.clickHandler}
        >
          <div className={styles.iconContainer}>
            {iconRenderer ||
              (icon && <div className={cn(styles.icon, icon)} />)}
            <div className={styles.text}>{text}</div>
          </div>
          {triggerRenderer ||
            (hasTrigger && (
              <div
                className={cn(styles.trigger, 'icon-cevron-down', {
                  [styles.isOpened]: isExpanded,
                })}
              />
            ))}
        </div>

        <div
          ref={this.content}
          onTransitionEnd={() => this.updateAfterTransition()}
          style={{ height: contentHeight }}
          className={cn(styles.content, { [content]: content })}
        >
          <div>{children}</div>
        </div>
      </div>
    );
  }
}
