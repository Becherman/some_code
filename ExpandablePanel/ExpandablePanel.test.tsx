import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { ExpandablePanel } from './ExpandablePanel';

describe('[Components] => <ExpandablePanel />', () => {
  let wrapper: ShallowWrapper;
  const props = {
    text: 'foo',
    onChange: jest.fn(),
    defaultExpanded: false,
    classes: {
      wrapper: 'foo',
      header: 'bar',
      content: 'baz',
    },
  };

  beforeEach(() => {
    wrapper = shallow(
      <ExpandablePanel {...props}>
        <div className="child"></div>
      </ExpandablePanel>,
    );

    props.onChange.mockClear();
  });

  it('renders without errors', () => {
    expect(wrapper.find('.wrapper').exists()).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  });

  it('adds a user classname to wrapper', () => {
    expect(wrapper.find('.wrapper').hasClass('foo')).toBe(true);
  });

  it('expanded by default if defaultExpanded prop was passed', () => {
    const wrapper = shallow(
      <ExpandablePanel {...props} defaultExpanded>
        <div className="child"></div>
      </ExpandablePanel>,
    );

    expect(wrapper.state('isExpanded')).toBe(true);
  });

  it('triggers onChange handler on every expand/collapse', () => {
    wrapper.find('.header').simulate('click');

    expect(props.onChange).toBeCalledWith(undefined, true);
  });

  it('becomes fully controllable component and ignore regular open/close mechanism', () => {
    const wrapper = shallow(
      <ExpandablePanel {...props} expanded>
        <div className="child"></div>
      </ExpandablePanel>,
    );

    expect(wrapper.state('isExpanded')).toBe(true);
    wrapper.find('.header').simulate('click');
    expect(wrapper.state('isExpanded')).toBe(true);
  });

  it('should consider defaultExpanded with more priority then expandede prop', () => {
    const wrapper = shallow(
      <ExpandablePanel {...props} expanded={false} defaultExpanded>
        <div className="child"></div>
      </ExpandablePanel>,
    );

    expect(wrapper.state('isExpanded')).toBe(true);
  });

  describe('Header', () => {
    it('renders header', () => {
      expect(wrapper.find('.header').exists()).toBeTruthy();
      expect(wrapper).toMatchSnapshot();
    });

    // Trigger
    it('renders trigger element by default', () => {
      expect(wrapper.find('.trigger').exists()).toBeTruthy();
    });

    it('renders passed triggerRenderer component instead of a basic trigger', () => {
      const wrapper = shallow(
        <ExpandablePanel
          {...props}
          triggerRenderer={<div className="triggerRenderer" />}
        >
          <div className="child"></div>
        </ExpandablePanel>,
      );

      expect(wrapper.find('.trigger').exists()).toBeFalsy();
      expect(wrapper.find('.triggerRenderer').exists()).toBeTruthy();
      expect(wrapper).toMatchSnapshot();
    });

    it('should add .isOpened class to trigger on click', () => {
      expect(wrapper.find('.trigger').hasClass('isOpened')).toBe(false);
      wrapper.find('.header').simulate('click');
      expect(wrapper.find('.trigger').hasClass('isOpened')).toBe(true);
      expect(wrapper).toMatchSnapshot();
    });

    it('renders trigger if hasTrigger prop was passed', () => {
      expect(wrapper.find('.trigger').exists()).toBe(true);
      wrapper.setProps({ hasTrigger: false });
      expect(wrapper.find('.trigger').exists()).toBe(false);
      expect(wrapper).toMatchSnapshot();
    });

    // Icon
    it('renders icon if icon prop was passed', () => {
      expect(wrapper.find('.icon').exists()).toBe(false);
      wrapper.setProps({ icon: 'foobar' });
      expect(wrapper.find('.icon').exists()).toBe(true);
      expect(wrapper).toMatchSnapshot();
    });

    it('renders passed iconRenderer component instead of a basic icon', () => {
      const wrapper = shallow(
        <ExpandablePanel
          {...props}
          triggerRenderer={<div className="iconRenderer" />}
        >
          <div className="child"></div>
        </ExpandablePanel>,
      );

      expect(wrapper.find('.icon').exists()).toBeFalsy();
      expect(wrapper.find('.iconRenderer').exists()).toBeTruthy();
      expect(wrapper).toMatchSnapshot();
    });

    //Text
    it('renders passed text prop', () => {
      expect(wrapper.find('.text').text()).toBe(props.text);
      expect(wrapper).toMatchSnapshot();
    });
  });

  // Content
  describe('Content', () => {
    it('renders content block', () => {
      expect(wrapper.find('.content').exists()).toBeTruthy();
      expect(wrapper).toMatchSnapshot();
    });

    it('renders passed child', () => {
      expect(wrapper.find('.child').exists()).toBeTruthy();
      expect(wrapper).toMatchSnapshot();
    });

    it('adds custom styles to content', () => {
      expect(wrapper.find('.content').hasClass('baz')).toBe(true);
    });
  });
});
