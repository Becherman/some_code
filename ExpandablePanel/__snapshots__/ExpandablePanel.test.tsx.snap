// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Components] => <ExpandablePanel /> Content renders content block 1`] = `
<div
  className="wrapper foo"
>
  <div
    className="header bar"
    onClick={[Function]}
  >
    <div
      className="iconContainer"
    >
      <div
        className="text"
      >
        foo
      </div>
    </div>
    <div
      className="trigger icon-cevron-down"
    />
  </div>
  <div
    className="content baz"
    onTransitionEnd={[Function]}
    style={
      Object {
        "height": 0,
      }
    }
  >
    <div>
      <div
        className="child"
      />
    </div>
  </div>
</div>
`;

exports[`[Components] => <ExpandablePanel /> Content renders passed child 1`] = `
<div
  className="wrapper foo"
>
  <div
    className="header bar"
    onClick={[Function]}
  >
    <div
      className="iconContainer"
    >
      <div
        className="text"
      >
        foo
      </div>
    </div>
    <div
      className="trigger icon-cevron-down"
    />
  </div>
  <div
    className="content baz"
    onTransitionEnd={[Function]}
    style={
      Object {
        "height": 0,
      }
    }
  >
    <div>
      <div
        className="child"
      />
    </div>
  </div>
</div>
`;

exports[`[Components] => <ExpandablePanel /> Header renders header 1`] = `
<div
  className="wrapper foo"
>
  <div
    className="header bar"
    onClick={[Function]}
  >
    <div
      className="iconContainer"
    >
      <div
        className="text"
      >
        foo
      </div>
    </div>
    <div
      className="trigger icon-cevron-down"
    />
  </div>
  <div
    className="content baz"
    onTransitionEnd={[Function]}
    style={
      Object {
        "height": 0,
      }
    }
  >
    <div>
      <div
        className="child"
      />
    </div>
  </div>
</div>
`;

exports[`[Components] => <ExpandablePanel /> Header renders icon if icon prop was passed 1`] = `
<div
  className="wrapper foo"
>
  <div
    className="header bar"
    onClick={[Function]}
  >
    <div
      className="iconContainer"
    >
      <div
        className="icon foobar"
      />
      <div
        className="text"
      >
        foo
      </div>
    </div>
    <div
      className="trigger icon-cevron-down"
    />
  </div>
  <div
    className="content baz"
    onTransitionEnd={[Function]}
    style={
      Object {
        "height": 0,
      }
    }
  >
    <div>
      <div
        className="child"
      />
    </div>
  </div>
</div>
`;

exports[`[Components] => <ExpandablePanel /> Header renders passed iconRenderer component instead of a basic icon 1`] = `
<div
  className="wrapper foo"
>
  <div
    className="header bar"
    onClick={[Function]}
  >
    <div
      className="iconContainer"
    >
      <div
        className="text"
      >
        foo
      </div>
    </div>
    <div
      className="iconRenderer"
    />
  </div>
  <div
    className="content baz"
    onTransitionEnd={[Function]}
    style={
      Object {
        "height": 0,
      }
    }
  >
    <div>
      <div
        className="child"
      />
    </div>
  </div>
</div>
`;

exports[`[Components] => <ExpandablePanel /> Header renders passed text prop 1`] = `
<div
  className="wrapper foo"
>
  <div
    className="header bar"
    onClick={[Function]}
  >
    <div
      className="iconContainer"
    >
      <div
        className="text"
      >
        foo
      </div>
    </div>
    <div
      className="trigger icon-cevron-down"
    />
  </div>
  <div
    className="content baz"
    onTransitionEnd={[Function]}
    style={
      Object {
        "height": 0,
      }
    }
  >
    <div>
      <div
        className="child"
      />
    </div>
  </div>
</div>
`;

exports[`[Components] => <ExpandablePanel /> Header renders passed triggerRenderer component instead of a basic trigger 1`] = `
<div
  className="wrapper foo"
>
  <div
    className="header bar"
    onClick={[Function]}
  >
    <div
      className="iconContainer"
    >
      <div
        className="text"
      >
        foo
      </div>
    </div>
    <div
      className="triggerRenderer"
    />
  </div>
  <div
    className="content baz"
    onTransitionEnd={[Function]}
    style={
      Object {
        "height": 0,
      }
    }
  >
    <div>
      <div
        className="child"
      />
    </div>
  </div>
</div>
`;

exports[`[Components] => <ExpandablePanel /> Header renders trigger if hasTrigger prop was passed 1`] = `
<div
  className="wrapper foo"
>
  <div
    className="header bar"
    onClick={[Function]}
  >
    <div
      className="iconContainer"
    >
      <div
        className="text"
      >
        foo
      </div>
    </div>
  </div>
  <div
    className="content baz"
    onTransitionEnd={[Function]}
    style={
      Object {
        "height": 0,
      }
    }
  >
    <div>
      <div
        className="child"
      />
    </div>
  </div>
</div>
`;

exports[`[Components] => <ExpandablePanel /> Header should add .isOpened class to trigger on click 1`] = `
<div
  className="wrapper foo"
>
  <div
    className="header bar"
    onClick={[Function]}
  >
    <div
      className="iconContainer"
    >
      <div
        className="text"
      >
        foo
      </div>
    </div>
    <div
      className="trigger icon-cevron-down isOpened"
    />
  </div>
  <div
    className="content baz"
    onTransitionEnd={[Function]}
    style={
      Object {
        "height": null,
      }
    }
  >
    <div>
      <div
        className="child"
      />
    </div>
  </div>
</div>
`;

exports[`[Components] => <ExpandablePanel /> renders without errors 1`] = `
<div
  className="wrapper foo"
>
  <div
    className="header bar"
    onClick={[Function]}
  >
    <div
      className="iconContainer"
    >
      <div
        className="text"
      >
        foo
      </div>
    </div>
    <div
      className="trigger icon-cevron-down"
    />
  </div>
  <div
    className="content baz"
    onTransitionEnd={[Function]}
    style={
      Object {
        "height": 0,
      }
    }
  >
    <div>
      <div
        className="child"
      />
    </div>
  </div>
</div>
`;
